src/ILLink.RoslynAnalyzer/RequiresAssemblyFilesAnalyzer.cs

```diff
@@ -134,17 +142,53 @@ public override void Initialize (AnalysisContext context)
 					CheckCalledMember (operationContext, methodSymbol, dangerousPatterns);
 				}, OperationKind.DelegateCreation);
 
+				static void CheckStaticConstructors (OperationAnalysisContext operationContext,
+					ImmutableArray<IMethodSymbol> constructors)
+				{
+					foreach (var constructor in constructors) {
+						if (constructor.Parameters.Length == 0 && constructor.HasAttribute (RequiresAssemblyFilesAttribute) && constructor.MethodKind == MethodKind.StaticConstructor) {
+							if (constructor.TryGetAttributeWithMessageOnCtor (FullyQualifiedRequiresAssemblyFilesAttribute, out AttributeData? requiresUnreferencedCode)) {
+								operationContext.ReportDiagnostic (Diagnostic.Create (
+									s_requiresAssemblyFilesRule,
+									operationContext.Operation.Syntax.GetLocation (),
+									constructor.ToString (),
+									(string) requiresUnreferencedCode!.ConstructorArguments[0].Value!,
+									requiresUnreferencedCode!.NamedArguments.FirstOrDefault (na => na.Key == "Url").Value.Value?.ToString ()));
+							}
+						}
+					}
+				}
+
 				static void CheckCalledMember (
 					OperationAnalysisContext operationContext,
 					ISymbol member,
 					ImmutableArray<ISymbol> dangerousPatterns)
 				{
+					// Find containing symbol
+					ISymbol? containingSymbol = null;
+					for (var current = operationContext.Operation;
+						 current is not null;
+						 current = current.Parent) {
+						if (current is ILocalFunctionOperation local) {
+							containingSymbol = local.Symbol;
+							break;
+						} else if (current is IAnonymousFunctionOperation lambda) {
+							containingSymbol = lambda.Symbol;
+							break;
+						} else if (current is IMethodBodyBaseOperation) {
+							break;
+						}
+					}
+					containingSymbol ??= operationContext.ContainingSymbol;
+
 					// Do not emit any diagnostic if caller is annotated with the attribute too.
-					if (operationContext.ContainingSymbol.HasAttribute (RequiresAssemblyFilesAttribute))
+					if (containingSymbol.HasAttribute (RequiresAssemblyFilesAttribute))
 						return;
 					// In case ContainingSymbol is a property accesor check also for RequiresAssemblyFilesAttribute in the associated property
-					if (operationContext.ContainingSymbol is IMethodSymbol containingSymbol && (containingSymbol.MethodKind == MethodKind.PropertyGet || containingSymbol.MethodKind == MethodKind.PropertySet)
-						&& containingSymbol.AssociatedSymbol!.HasAttribute (RequiresAssemblyFilesAttribute)) {
+					if (containingSymbol is IMethodSymbol methodSymbol &&
+						(methodSymbol.MethodKind == MethodKind.PropertyGet || methodSymbol.MethodKind == MethodKind.PropertySet ||
+						methodSymbol.MethodKind == MethodKind.EventAdd || methodSymbol.MethodKind == MethodKind.EventRemove)
+						&& methodSymbol.AssociatedSymbol!.HasAttribute (RequiresAssemblyFilesAttribute)) {
```

Comment:
I think you could probably just say `methodSymbol.AssociatedSymbol?.HasAttribute(RequiresAssemblyFilesAttribute)` and it would work fine.